### ISSUE TITLE
But in R/lmer.R: all models must be fit to the same data object

### ISSUE THREAD

**Issue by mdg583** [lme4/lme4#622]
Hi,
I think I've found a bug in R/lmer.R, lines 545 - 548:

```
        calls <- lapply(mods, getCall)
        data <- lapply(calls, `[[`, "data")
        if(!all(vapply(data, identical, NA, data[[1]])))
            stop("all models must be fit to the same data object")
```

I believe this is comparing that datasets have the same names, rather than that they are identical datasets.

Here is a simple example:
```
library(lme4)

data(mtcars)

cars1 <- na.omit(mtcars)
cars2 <- na.omit(mtcars)
m1 <- lmer(mpg ~ hp+(1|gear),data=cars1)
m2 <- lmer(mpg ~ (1|gear),data=cars2)
anova(m1,m2)
# error: all models must be fit to the same data object, even though datasets are identical

cars1$gear <- cars1$gear+1
m2 <- lmer(mpg ~ (1|gear),data=cars1)
anova(m1,m2)
# Datasets are not the same, but no error raised
```

**Comment by bbolker** [comment]
I agree that this is questionable, although it is working as originally intended. I wouldn't necessarily *object* to testing whether the data sets are the same, but here are some issues (= reasons why it might be better to leave the behaviour as it is but document it better ...)

1. although admittedly weird stuff can happen in the course of a data analysis (especially if people have unstructured/interactive workflows), the two cases above (identical data sets with different names; data sets with the same name that happen to be different) seem like edge cases/user errors. The latter seems more plausible.
2. There could be some computational burden to testing whether large data sets are identical?
3. How picky would one want to be when testing identity? Use `all.equal()` or `identical()`? What about trivial/unimportant differences (e.g. differences in row names)?

Should there be an option to ignore/convert to a warning (similar to the `*lmerControl()` check options)?

**Comment by mdg583** [comment]
My use case is as follows:

I was creating a function to generate some info comparing nested models. I wanted to be able to make this comparison even if the 'outer' model had some additional missing values (in the extra terms). I wanted to generate an anova Chi-square p-value.

I did this by re-fitting the nested model with observations removed which had the missing values in the outer model. This worked at first, but eventually stopped working with the "all models must be fit to the same data object" error, even though the dataframes seemed to be the same. After a lot of hunting I eventually saw that they simply needed to have the same name in the model fitting call.

(I'm not 100% sure of how valid or useful it is to compare using a Chi-square p-value after removing missing terms)

I'm wondering if this check is meant to enforce that the user compares models fit to the same data? I guess it would have been nice if this was clearer in the error message.

**Comment by bbolker** [comment]
In general (unless you're doing something a little tricky with transforming predictors on the fly), `model_data <- model.frame(full_model_formula, data=original_data)` should extract the data that has complete cases for the full model (i.e., no missing values for any of the predictors).

Any chance of a small reproducible example?

A necessary-but-not-sufficient check is that `nobs()` is the same for all models.

Yes, the message means that all models need to be fitted to the same data (it's true that this doesn't need to be the same object, but most generally it would be ...) How would you rephrase the error message to make it clearer?

**Comment by mdg583** [comment]
I feel like if this were a warning rather than an error message, it would be easier to understand what is meant. Not having the name for the dataset suggests the user may not be fitting the models to the same dataset, but doesn't mean that is the case necessarily.
